### ISSUE TITLE
SC 25 | MLP-Offload: Multi-Level, Multi-Path Offloading for LLM Pre-training to Break the GPU Memory Wall

### ISSUE THREAD

**Issue by Dszdsxc0405** [AI-Infra-Team/awesome-papers#140]
## Paper Infomation

Produced in the United States

**Title:** MLP-Offload: Multi-Level, Multi-Path Offloading for LLM Pre-training to Break the GPU Memory Wall
**Authors:** Avinash Maurya,  M. Mustafa Rafique, Franck Cappello, Bogdan Nicolae
**Institudes:**  
1.Argonne National Laboratory Lemont, IL, USA [Argonne National Laboratory Homepage | Argonne National Laboratory](https://www.anl.gov/)
2.Rochester Institute of Technology Rochester, NY, USA
**Paper:** [MLP-Offload: Multi-Level, Multi-Path Offloading for LLM Pre-training to Break the GPU Memory Wall | Proceedings of the International Conference for High Performance Computing, Networking, Storage and Analysis](https://dl.acm.org/doi/10.1145/3712285.3759864)
**Github Note Link:** [mauryaavinash95/MLP-Offload: Software Artifacts for SC'25 paper MLP-Offload](https://github.com/mauryaavinash95/MLP-Offload/tree/master)

### Background
As the scale of large language models grows, a fundamental contradiction has become increasingly prominent: **the growth rate of data structures such as model parameters, activations, and optimizer states (increasing approximately 450 times every two years) has far surpassed the improvement speed of GPU memory capacity (approximately doubling every two years)**, a phenomenon known as the **"GPU Memory Wall."**

To efficiently train ultra-large models on limited GPUs, existing mainstream frameworks like DeepSpeed and PyTorch FSDP adopt a combination of technologies, including **Zero Redundancy Optimizer (ZeRO)**, **mixed-precision training**, and **multi-tier offloading**.
However, this introduces new performance bottlenecks: the I/O bandwidth of storage media is significantly lower than that of GPU memory:
> "When pre-training a 20-billion parameter LLaMA model without offloading (with model parameters and optimizer states residing in the aggregated GPU memory of a single node), the average iteration duration is 0.4 seconds. Offloading the optimizer states used during the update phase to host memory increases the average iteration duration to 3.7 seconds. Further offloading to node-local NVMe drives increases the iteration duration to 67 seconds (§3.1)."

### Findings
Through an in-depth performance analysis of state-of-the-art offloading technologies represented by DeepSpeed ZeRO-3, the paper finds that:
1.  **Training iterations are dominated by the update phase**: When using SSD offloading, the update phase can account for 89% of a single iteration's time (taking a 40-billion parameter model as an example).
2.  **The update phase is I/O-bound**: When subgroups are offloaded to SSDs, 99% of the update phase time is spent on SSD I/O, even with asynchronous computation overlap.
3.  **Remote storage bandwidth is underutilized**: Remote storage, such as Parallel File Systems (PFS) common in HPC systems, typically remains idle during training (except for checkpoints). Their aggregated bandwidth could supplement local NVMe.
4.  **Concurrent offloading intensifies I/O contention**: Multiple GPU processes concurrently accessing shared local NVMe drives lead to saturated I/O bandwidth and increased latency, becoming a bottleneck during scaling.
<div align="center"> 
    <img width="650" height="297" alt="Image" src="https://github.com/user-attachments/assets/cd2a5609-342f-4bbc-af54-a8d9bcd39ddd" />
</div>

### Contributions
- Based on these insights, the authors propose **MLP-Offload** (Multi-Level, Multi-Path Offloading). 
- Its core idea is: **By creating a unified "virtual" storage layer that intelligently integrates heterogeneous storage resources such as local NVMe and remote parallel file systems, leveraging multi-path parallel I/O to aggregate bandwidth, and combining caching optimizations and concurrency control strategies to maximize the overlap of I/O and computation, thereby significantly alleviating the I/O bottleneck in the update phase.**

## Method

1.  **Unified Multi-Level, Multi-Path Asynchronous Offloading (Virtual Tier)**:
    
    -   **Idea**: Abstracts heterogeneous physical storage (local NVMe, remote PFS, object storage) into a unified "virtual third tier." Instead of simply dumping all data to the fastest storage, the system intelligently routes data subgroups to different paths based on the **real-time bandwidth** of each physical path.
        
    -   **Method**: Based on a performance model (see Formula 1 below), allocates the number of subgroups proportionally to bandwidth. For example, if local NVMe bandwidth is twice that of remote PFS, approximately two-thirds of the subgroups are stored on NVMe and one-third on PFS. This ensures the time to fetch subgroups in parallel from both paths is roughly equal, achieving I/O load balancing.
        
    -   **Advantage**: **Aggregates the bandwidth of multiple storage devices**, preventing any single slow device from becoming a bottleneck. Fully utilizes the PFS bandwidth in HPC systems that is often idle during training tasks, representing a deep exploitation of the potential of existing storage hierarchies.

2.  **Optimized Virtual Tier Concurrency Control**:
    
    -   **Idea**: To prevent bandwidth contention and latency spikes caused by multiple GPU processes on the same node simultaneously reading from/writing to the same physical device (e.g., a single NVMe SSD), a coordination mechanism is needed.
        
    -   **Method**: Implements a lightweight "exclusive-access" lock at the node level. For each physical storage device, only one process is allowed to perform I/O at any given time. This process can saturate the device's bandwidth, while other processes either perform computations or access other available storage devices.
        
    -   **Advantage**: **Eliminates inter-process I/O interference**, ensuring low latency and high bandwidth for individual I/O operations. This intra-node coordination achieves natural, global interleaving of I/O and computation.

3.  **Cache-Friendly Model Subgroup Processing Order**:
    
    -   **Idea**: Since the processing of subgroups by the optimizer update (e.g., Adam) is mutually independent, the order does not affect the result. This property is leveraged to adjust the processing order to maximize host memory cache hit rates.
        
    -   **Method**: Alternates between processing subgroup IDs in ascending and descending order across consecutive iterations. For instance, if iteration 1 processes in ascending order, high-ID subgroups remain in host memory last. If iteration 2 processes in descending order, it first processes these high-ID subgroups still cached in memory, avoiding I/O.
        
    -   **Advantage**: **Significantly reduces "thrashing" caused by insufficient cache capacity**, where subgroups just fetched into memory are quickly evicted. Figure 10 in the paper shows that this method allows more optimizer state subgroups to reside in faster host memory, thereby reducing the frequency of data fetches from slower storage.
<div align="center">
    <img width="280" height="220" alt="Image" src="https://github.com/user-attachments/assets/bafa0e08-2248-4d16-aea8-b22609174688" />
</div>

4. **Lazy In-Place Mixed-Precision Gradient Conversion**:
    
    -   **Idea**: The traditional method converts FP16 gradients to FP32 and offloads them immediately after the backward pass. This increases the amount of data that needs to be transferred (FP32 is twice the size of FP16), and the conversion operation itself can block the backward pass.
        
    -   **Method**: During the backward phase, only FP16 gradients are kept in the accumulation buffer in host memory. During the update phase, only the optimizer state subgroups (without gradients) are retrieved from slow storage. Then, the corresponding FP16 gradients are converted in-place to FP32 on the CPU in real-time for use in the update.
        
    -   **Advantage**: **Reduces the data volume for slow storage I/O** (no need to read/write the large FP32 gradients). Simultaneously overlaps the compute-intensive precision conversion operation with the read operation from slow storage. Since the CPU's conversion throughput far exceeds storage I/O bandwidth, the additional overhead is nearly negligible.
<div align="center">
    <img width="850" height="453" alt="Image" src="https://github.com/user-attachments/assets/f5d36760-341d-44f1-ac71-65b3b7e61a94" />
</div>

## Experiments
- **Experimental Setup**:
<div align="center">
    <img width="408" height="183" alt="Image" src="https://github.com/user-attachments/assets/3ff1134e-45a3-4544-b93c-490d35639abd" />
</div>

- **Baseline for Comparison**: DeepSpeed ZeRO-3 (with the DeepNVMe asynchronous offloading engine).

- **Models Evaluated**: A series of model configurations ranging from 40 billion to 280 billion parameters were used (based on the architectures of real models like LLaMA and Bloom), covering small, medium, and large scales.
<div align="center">
    <img width="400" height="144" alt="Image" src="https://github.com/user-attachments/assets/e8fe2788-5e94-4d5e-b742-583e82024605" />
</div>

- **Key Metrics**:
    1. Average single training iteration time (broken down into forward, backward, and update phases).
    2. Update throughput (millions of parameters updated per second).
    3. Effective I/O throughput (actual bandwidth for reading and writing subgroups from/to storage).
    4. Distribution of optimizer states across different storage tiers.

### **Key Results Analysis**:

1. **Model Scaling (§4.2)**:
    - As model size increased from 40B to 120B, offloading pressure increased. **MLP-Offload significantly reduced iteration time across all scales**, being 2.7x faster overall than DeepSpeed. Simultaneously, MLP-Offload achieved 1.8-2.4x higher update throughput than DeepSpeed and remained relatively stable.
<div align="center">
    <img width="551" height="232" alt="Image" src="https://github.com/user-attachments/assets/0329501b-a980-4ad6-85c4-8038016994b3" />
</div>

2. **I/O and Storage Tier Load (§4.3)**:
    - Figure 9 shows a key finding: The actual effective I/O throughput of DeepSpeed ZeRO-3 (~3.2 GB/s) was far lower than the peak write bandwidth of the testbed's local NVMe (5.3 GB/s), **demonstrating that concurrent competition caused severe underutilization of bandwidth**. In contrast, by introducing PFS as a second path, MLP-Offload achieved an aggregated I/O throughput of ~8.3 GB/s, which is **2.6x** that of DeepSpeed.
<div align="center">
    <img width="284" height="224" alt="Image" src="https://github.com/user-attachments/assets/2c581139-96bc-4206-b33e-4a2500e9ba8d" />
</div>

3. **Weak Scaling (§4.4)**:
    - The model scale and the number of GPUs were increased proportionally (from 1 node/40B to 8 nodes/280B). Figure 11 shows that even when scaling to 32 A100 GPUs, **MLP-Offload maintained roughly a 2x advantage in iteration speed**. This indicates that its optimizations remain effective in distributed environments, and communication overhead did not negate the benefits gained from I/O optimization.
    - The paper also provides a meaningful "cost-performance" comparison: Training a 70B model required ~80 A100 GPUs without offloading (24 sec/iter); using DeepSpeed offloading required only 8 A100 GPUs but took 168 sec/iter (7x slower); while MLP-Offload, also using 8 A100 GPUs, took only 115 sec/iter (4.8x slower). **MLP-Offload saved 10x in GPUs while incurring only about a 5x performance penalty, achieving a 2x improvement in cost-performance.**
<div align="center">
    <img width="324" height="256" alt="Image" src="https://github.com/user-attachments/assets/478a4a9f-b987-4ca4-a1c5-17a7b2cb3c5a" />
</div>

4. **Ablation Study (§4.6)**:
    (Omitted)
